cfgrammar/src/lib/yacc/ast.rs

```diff
@@ -202,8 +230,101 @@ impl GrammarAST {
+                    if self.get_rule(sym_name).is_none() {
+                        return Err(YaccGrammarError {
+                            kind: YaccGrammarErrorKind::UnknownRuleRef(sym_name.clone()),
+                            spans: vec![*sym_span],
+                        });
+                    }
+                }
+                Symbol::Token(sym_name, sym_span) => {
+                    if !self.has_token(sym_name) {
+                        return Err(YaccGrammarError {
+                            kind: YaccGrammarErrorKind::UnknownToken(sym_name.clone()),
+                            spans: vec![*sym_span],
+                        });
+                    }
+                }
+            }
+        }
         Ok(())
     }
+
+    /// Returns tokens, and rules unreachable from the start rule
+    /// by index into the `ast.tokens` and `ast.rules` fields.
+    pub(crate) fn unused_symbols(&self) -> impl Iterator<Item = SymbolIdx> + '_ {
+        let start_rule_name = self.start.as_ref().map(|(name, _)| name.clone());
+        let start_rule = self
+            .rules
+            .iter()
+            .find(|(rule_name, _)| start_rule_name.as_ref() == Some(rule_name));
+        let mut seen_rules = HashSet::new();
+        let mut seen_tokens = HashSet::new();
+        let mut expected_unused_tokens = HashSet::new();
+        let mut expected_unused_rules = HashSet::new();
+        for sym in &self.expect_unused {
+            match sym {
+                Symbol::Rule(sym_name, _) => {
+                    expected_unused_rules.insert(sym_name);
+                }
+                Symbol::Token(sym_name, _) => {
+                    expected_unused_tokens.insert(sym_name);
+                }
+            }
+        }
+        if let Some(implicit_tokens) = self.implicit_tokens.as_ref() {
+            expected_unused_tokens.extend(implicit_tokens.keys())
+        }
+        if let Some((start_name, start_rule)) = start_rule {
+            let mut todo = Vec::new();
+            todo.extend(start_rule.pidxs.iter().copied());
+            seen_rules.insert(start_name);
+
+            while let Some(pidx) = todo.pop() {
+                let prod = &self.prods[pidx];
+                for sym in &prod.symbols {
+                    match sym {
+                        Symbol::Rule(name, _) => {
+                            if seen_rules.insert(name) {
+                                if let Some(rule) = self.rules.get(name) {
+                                    todo.extend(&rule.pidxs);
+                                }
+                            }
+                        }
+                        Symbol::Token(name, _) => {
+                            seen_tokens.insert(name);
+                        }
+                    };
+                }
+            }
+        }
+        self.rules
+            .iter()
+            .enumerate()
+            .filter_map(move |(rule_id, (rule_name, _))| {
```

Comment:
Were we able to use `difference` here or did it not work?

Reply:
Yeah, I never could get it to work due to the type differences between `IndexMap`/`IndexSet` of `self.rules`, `self.tokens` and the `HashMap`s of the `expected` and `seen` tables.

Reply:
OK.